doc: dev model: Document NCS redistribution (#9773)
nrfconnect/sdk-nrf: doc/nrf/dm_ncs_distro.rst

.. _dm_ncs_distro:

Redistributing the |NCS|
########################

.. contents::
   :local:
   :depth: 2

This section is intended for users that would like to re-distribute the |NCS| with additions or changes to it.
It can be useful if you are a software or hardware vendor that would like to augment the |NCS| with additional functionality, and still benefit from the infrastructure it provides.

The contents of this section is not comprehensive.
You can rather consider it as suggestions for extending the |NCS|.
You can use other mechanisms as long as they comply with the text in the :file:`LICENSE` file located in the root of every Git repository in the |NCS|.
If you need any help, reach out to Nordic Semiconductor.

Getting added to the main manifest
**********************************

The simplest way to become part of the |NCS| is to be included in its official `west manifest file`_.
This way, your own :ref:`module repository <zephyr:modules>` will automatically be part of the |NCS| official distribution and will be included in all releases.

> This solution is reserved for selected vendors and requires a commercial agreement with Nordic Semiconductor.

Reviewer: I am new to this NCS and following might be covered: Does the commercial agreement define testing and compatibility responsibilities?  Is Nordic some how responsible on verifying compatibility or does the redistributor take the responsibility?

If this page is for the contributors, about "how to become a redistributor" or "How to be part of nRF Connect SDK delivery". Do we have some page for the users describing that SDK contains parts of other contributors... what are nordic own and what parts are 3rd party?

Author: >I am new to this NCS and following might be covered: Does the commercial agreement define testing and compatibility responsibilities? Is Nordic some how responsible on verifying compatibility or does the redistributor take the responsibility?

Very good question. There is no generic answer to this, it really depends on each individual agreement with Nordic how testing, compatibility and maintenance are handled.

>If this page is for the contributors, about "how to become a redistributor" or "How to be part of nRF Connect SDK delivery

This page is for those who want to redistribute nRF Connect SDK, not just use it to develop a product.

>Do we have some page for the users describing that SDK contains parts of other contributors... what are nordic own and what parts are 3rd party?

We don't specifically, but it's relatively simple to find out based on the [manifest](https://github.com/nrfconnect/sdk-nrf/blob/main/west.yml) or the [repo revision table](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/introduction.html#ncs-repository-revisions). For each repository you can see whether it's Nordic-developed or 3rd-party code by looking at the license and description.

Contact your local representative if you wish to be considered for inclusion.

Some examples of this type of integration are:

* :ref:`Memfault <mod_memfault>`.
  This functionality is present through the addition of a `west project in the nRF Connect SDK manifest <Memfault entry in the manifest_>`_.
  The `repository provided by Memfault <Memfault firmware SDK_>`_ is public.
  By default, |NCS| users will obtain a copy of it (as well as the integration code necessary for |NCS| based applications) when :ref:`getting the code <cloning_the_repositories>`.

* ANT protocol support.
  The code and documentation of this protocol is confidential, and access to the private module repository requires the GitHub user to be added to it.
  This is why the west project is disabled by default, using the :ref:`project groups <zephyr:west-manifest-groups>` feature of west.
  This means that the ANT entry in the manifest is disabled by default through its presence in the `manifest group filter`_.

>   You will not get a copy of this repository when :ref:`getting the code <cloning_the_repositories>`, and instead, need to enable the repository first using west itself to fetch it locally::

Reviewer 2: double :

Author: That's intended, to get monospace in the line below.

>     west config manifest.group-filter +ant

Reviewer 2: 2 queries:
1) How does someone enable this from ncs toolchain manager or would it be the same?
2) Is it worth linking to documentation (i.e. git or github) about how a user sets up authentication for the protected repos?

Author: > How does someone enable this from ncs toolchain manager or would it be the same?

It would be exactly the same, from the command-line.

> Is it worth linking to documentation (i.e. git or github) about how a user sets up authentication for the protected repos?

This is outside the scope of this section in my opinion, given that this is about redistribution. I think we should add a chapter somewhere in the NCS Getting Started doc mentioning this, will work on it on a separate PR.

    west update

Providing a Zephyr module
*************************

This approach is essentially identical to the previous one.
However, the :ref:`module repository <zephyr:modules>` you create will not be part of the |NCS| official distribution by default.
Instead, you will need to host your own documentation explaining how to add the module you make available to user's :ref:`own manifest <dm_workflow_4>`.

An example of this approach is the `Golioth Zephyr SDK`_, which is set up as a `Zephyr module <Golioth module.yml_>`_ that a user can add to their own manifest.

Providing a west manifest
*************************

Instead of just providing a module repository like in the previous approach, you can also include a west manifest that imports the |NCS| in it, see :ref:`dm_workflow_4`.
With this approach, you can import your manifest directly.
It will then in turn import the |NCS| `west manifest file`_.

Just like in the previous approach, Golioth provides an example of such a west manifest, `west-ncs.yml <Golioth west-ncs.yml_>`_ along with a `guide document <Golioth README nRF Connect SDK_>`_.

Forking the |NCS|
*****************

> You can always :ref:`fork all or parts of the nRF Connect SDK <dm-wf-fork>` and then make all the changes required by your features.

Reviewer 2: Had to do this in the past to change the mcuboot module, needed both mcuboot and sdk-nrf forked (as if you just tried to skip getting mcuboot in a top level west manifest and added your own, a fork of sdk-mcuboot, it did not work) - is it worth mentioning something like this an an example of when a fork would be needed?

Author: I think this would be going into too much detail for the scope of this section. Instead this info would belong somewhere like here:
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/dm_adding_code.html#workflow-3-application-in-a-fork-of-sdk-nrf

.. _dm_ncs_distro:

Redistributing NCS
##################

.. contents::
   :local:
   :depth: 2

This section is intended for users that would like to re-distribute the |NCS|
with additions or changes to it.  This can be useful if you are a software or
hardware vendor that would like to augment the |NCS| with additional

> functionality, while at the same time still benfitting from the infrastructure

Reviewer 3: nit. `benfitting` should be `benefiting`

Author: Thanks!

  the code <cloning_the_repositories>`.

* ANT protocol support. In this particular case the code and documentation of
  this protocol is confidential, and in fact access to the private module repository
  requires the GitHub user to be added to it. This is why the west project is
  disabled by default, using the :ref:`project groups
  <zephyr:west-manifest-groups>` feature of west. This means that the ANT entry
  in the manifest is disabled by default via its presence in the `group
  filter <manifest group filter>`_. Users will not get a copy of this repository
  when :ref:`getting the code <cloning_the_repositories>`, and will instead need
  to enable the repo first using west itself to fetch it locally::

    west config manifest.group-filter +ant
    west update

Providing a Zephyr module
*************************

This approach is essentially identical to the previous one, but with the
difference that the :ref:`module repository <zephyr:modules>` you create will
not be part of the |NCS| official distribution by default. Instead, you will
need to host your own documentation explaining users how to add the module you

> make available to their :ref:`own manifest <dm_workflow_4>`.

Reviewer 3: Optional, but we have documentation as an example in our [README](https://github.com/golioth/golioth-zephyr-sdk#using-with-nrf-connect-sdk) and [documentation site](https://docs.golioth.io/hardware/nrf91/zephyr-quickstart/set-up-zephyr).

Author: Added, thanks!

> An example of this approach is the `Golioth firmware SDK`_, which is set up as

Reviewer 3: Should be `Golioth Zephyr SDK`.

.. _dm_ncs_distro:

> Redistributing NCS
> ##################

Reviewer 4: ```suggestion
Redistributing |NCS|
####################
```

Reviewer 5: Oh, missed that.

.. contents::
   :local:
   :depth: 2

This section is intended for users that would like to re-distribute the |NCS|
with additions or changes to it.  This can be useful if you are a software or
hardware vendor that would like to augment the |NCS| with additional
functionality, while at the same time still benefiting from the infrastructure
it provides.

The contents of this section are not comprehensive, they are intended as
suggestions for those wanting to extend the |NCS| but other mechanisms may be
used as long as they comply with the text in the :file:`LICENSE` file that you
can find at the root of every Git repository in the |NCS|. In case of doubt

> please reach out to Nordic Semiconductor.

Author: I will modify "the contents" for "the content"

Getting added to the main manifest
**********************************

The simplest way (for your users or customers) to become part of the |NCS| is to
be included in its official `west manifest file`_. In this way your own
:ref:`module repository <zephyr:modules>` will automatically be part of the
|NCS| official distribution and will be included in all releases.

This solution is reserved for selected vendors and requires a commercial
agreement with Nordic Semiconductor. Please contact your local representative if
you'd like to be considered for inclusion.

Some examples of this type of integration are:

* :ref:`Memfault <mod_memfault>`. This functionality is present via the addition

>   of a `west project in the |NCS| manifest <Memfault entry in the manifest>`_.

Author: @peknis this type of link doesn't seem to work with Sphinx. Can you suggest an alternative?

Reviewer 5: You were just missing the underscore as fixed in my suggestion.

Reviewer 4: It should be \`west project in the nRF Connect SDK manifest \<Memfault entry in the manifest_\>\`_. I don't think the shortcut will work here.

Author: Actually it does! I checked and @peknis is right, just adding the underscore fixes i!

Author: Funnily... it works sometimes. This particular one works fine, but others don't seem to. Very strange.

Author: Actually scratch that. It seems to work fine now.

>   of a `west project in the |NCS| manifest <Memfault entry in the manifest>`_.

Reviewer 4: It should be \`west project in the nRF Connect SDK manifest \<Memfault entry in the manifest_\>\`_. I don't think the shortcut will work here.

  The `repository provided by Memfault <Memfault firmware SDK>`_ is public,
  and so |NCS| users will obtain a copy of it (as well as the integration code

>   necessary to be used in |NCS|-based aplicaitons) by default when :ref:`getting

Reviewer 6: Typo: `aplicaitons`

Author: Thanks! will fix

  the code <cloning_the_repositories>`.

> * ANT protocol support. In this particular case the code and documentation of
>   this protocol is confidential, and in fact access to the private module repository
>   requires the GitHub user to be added to it. This is why the west project is
>   disabled by default, using the :ref:`project groups
>   <zephyr:west-manifest-groups>` feature of west. This means that the ANT entry
>   in the manifest is disabled by default via its presence in the `group
>   filter <manifest group filter>`_. Users will not get a copy of this repository
>   when :ref:`getting the code <cloning_the_repositories>`, and will instead need
>   to enable the repo first using west itself to fetch it locally::

Reviewer 5: ```suggestion
* ANT protocol support.
  The code and documentation of this protocol is confidential, and access to the private module repository requires the GitHub user to be added to it.
  This is why the west project is disabled by default, using the :ref:`project groups <zephyr:west-manifest-groups>` feature of west.
  This means that the ANT entry in the manifest is disabled by default through its presence in the `group filter <manifest group filter_>`_.
  You will not get a copy of this repository when :ref:`getting the code <cloning_the_repositories>`, and instead, need to enable the repository first using west itself to fetch it locally::
```

    west config manifest.group-filter +ant
    west update

Providing a Zephyr module
*************************

> This approach is essentially identical to the previous one, but with the
> difference that the :ref:`module repository <zephyr:modules>` you create will
> not be part of the |NCS| official distribution by default. Instead, you will
> need to host your own documentation explaining users how to add the module you
> make available to their :ref:`own manifest <dm_workflow_4>`.

Reviewer 5: ```suggestion
This approach is essentially identical to the previous one.
However, the :ref:`module repository <zephyr:modules>` you create will not be part of the |NCS| official distribution by default.
Instead, you will need to host your own documentation explaining how to add the module you make available to user's :ref:`own manifest <dm_workflow_4>`.
```

> An example of this approach is the `Golioth Zephyr SDK`_, which is set up as
> a `Zephyr module <Golioth module.yml>`_ that any user can add to their own
> manifest.

Reviewer 5: ```suggestion
An example of this approach is the `Golioth Zephyr SDK`_, which is set up as a `Zephyr module <Golioth module.yml_>`_ that a user can add to their own manifest.
```

Providing a west manifest
*************************

> Instead of just providing a module repository like in the previous approach, you
> can choose to also include a west manifest that imports the |NCS| in it, very
> much like a :ref:`user would do <dm_workflow_4>`. With this approach the user
> can then import your manifest directly, which will then in turn import the |NCS|
> `west manifest file`_.

Reviewer 5: ```suggestion
Instead of just providing a module repository like in the previous approach, you can also include a west manifest that imports the |NCS| in it, see :ref:`dm_workflow_4`.
With this approach, you can then import your manifest directly.
It will then in turn import the |NCS| `west manifest file`_.
```

> Just like in the previous approach, Golioth provides an example of such a west
> manifest, `west-ncs.yml <Golioth west-ncs.yml>`_ along with a `small guide
> <Golioth README nRF Connect SDK>`_ for users.

Reviewer 5: ```suggestion
Just like in the previous approach, Golioth provides an example of such a west manifest, `west-ncs.yml <Golioth west-ncs.yml>`_ along with a `guide document <Golioth README nRF Connect SDK>`_.
```

> Forking NCS
> ***********

Reviewer 5: ```suggestion
Forking |NCS|
*************
```

> Another approach would be to just :ref:`fork all or parts of the |NCS|
> <dm-wf-fork>` and then make all the changes required by your features.

Reviewer 5: ```suggestion
You can always :ref:`fork all or parts of the |NCS| <dm-wf-fork>` and then make all the changes required by your features.
```
